sys/vm/vm_fault.c

```diff
@@ -2100,9 +2145,12 @@ vm_fault_copy_entry(vm_map_t dst_map, vm_map_t src_map,
 		 * backing pages.
 		 */
 		if (vm_page_all_valid(dst_m)) {
+			vm_prot_t realprot =
+			    VM_OBJECT_MASK_CAP_PROT(dst_object, prot);
+			realprot = vm_page_mask_cap_prot(dst_m, realprot);
+
 			VM_OBJECT_ASSERT_CAP(dst_object, prot);
```

Comment:
I would perhaps add the `realprot` logic below this assertion so that you validate `prot` first before using it as an input to compute `realprot`.